File: apps/api/app/schemas/campaign.py
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel

from .base import BaseSchema, MoneySerializerMixin, TimestampMixin
from .invoice import InvoiceSummary
from .line_item import LineItemInCampaign


class CampaignBase(BaseSchema):
    """Base Campaign fields"""

    id: int
    name: str


class CampaignListItem(MoneySerializerMixin, CampaignBase):
    """GET /api/v1/campaigns list item"""

    total_booked: Decimal
    total_actual: Decimal
    total_billable: Decimal
    line_items_count: int
    invoice_id: int | None


class CampaignDetail(MoneySerializerMixin, CampaignBase, TimestampMixin):
    """GET /api/v1/campaigns/{id} detail"""

    line_items: list[LineItemInCampaign]
    invoice_summary: InvoiceSummary | None = None


class CampaignListResponse(BaseModel):
    """GET /api/v1/campaigns response"""

    campaigns: list[CampaignListItem]
    total: int
    limit: int
    offset: int

File: apps/web/src/types/campaign.ts
export interface LineItemInCampaign {
  id: number;
  campaign_id: number;
  name: string;
  booked_amount: string;
}

export interface InvoiceSummary {
  id: number;
  campaign_id: number;
  total_actual: string;
  total_adjustments: string;
  total_billable: string;
  line_items_count: number;
}

export interface CampaignListItem {
  id: number;
  name: string;
  total_booked: string;
  total_actual: string;
  total_billable: string;
  line_items_count: number;
  invoice_id: number | null;
}

export interface CampaignDetail {
  id: number;
  name: string;
  created_at: string;
  updated_at: string;
  line_items: LineItemInCampaign[];
  invoice_summary: InvoiceSummary | null;
}

export interface CampaignListResponse {
  campaigns: CampaignListItem[];
  total: number;
  limit: number;
  offset: number;
}
